atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SerializableTransactionManagerTest.java

```diff
@@ -46,76 +48,100 @@
     private TimelockService mockTimelockService = mock(TimelockService.class);
     private Cleaner mockCleaner = mock(Cleaner.class);
     private AsyncInitializer mockInitializer = mock(AsyncInitializer.class);
-    private Runnable mockCallback = mock(Runnable.class);
+    private Callback mockCallback = mock(Callback.class);
 
     private SerializableTransactionManager manager;
 
     @Before
     public void setUp() {
-        when(mockKvs.isInitialized()).thenReturn(false);
+        setInitializationStatus(false, false, false, false);
         manager = getManagerWithCallback(true, mockCallback);
+    }
 
-        when(mockTimelockService.isInitialized()).thenReturn(true);
-        when(mockCleaner.isInitialized()).thenReturn(true);
-        when(mockInitializer.isInitialized()).thenReturn(true);
+    @Test
+    public void uninitializedTransactionManagerThrowsNotInitializedException() {
+        assertNotInitializedWithinTwoSeconds();
+        Assertions.assertThatThrownBy(() -> manager.runTaskWithRetry($  -> null))
+                .isInstanceOf(NotInitializedException.class);
     }
 
     @Test
-    public void isInitializedWhenPrerequisitesAreInitialized() {
-        when(mockKvs.isInitialized()).thenReturn(true);
+    public void isInitializedAndCallbackHasRunWhenPrerequisitesAreInitialized() {
+        setInitializationStatus(true, true, true, true);
         Awaitility.waitAtMost(2L, TimeUnit.SECONDS).until(() -> manager.isInitialized());
-        verify(mockCallback).run();
+        verify(mockCallback).runWithRetry();
     }
 
     @Test
-    public void closingPreventsCallback() {
-        manager.close();
-        when(mockKvs.isInitialized()).thenReturn(true);
+    public void switchBackToUninitializedWhenPrerequisitesSwitchToUninitialized() {
+        setInitializationStatus(true, true, true, true);
+        Awaitility.waitAtMost(2L, TimeUnit.SECONDS).until(() -> manager.isInitialized());
+
+        setInitializationStatus(false, false, false, false);
         assertNotInitializedWithinTwoSeconds();
-        verify(mockCallback, never()).run();
+        Assertions.assertThatThrownBy(() -> manager.runTaskWithRetry($  -> null))
+                .isInstanceOf(NotInitializedException.class);
+    }
+
+    @Test
+    public void callbackRunsOnlyOnceAsInitializationStatusChanges() {
+        setInitializationStatus(true, true, true, true);
+        Awaitility.waitAtMost(2L, TimeUnit.SECONDS).until(() -> manager.isInitialized());
+
+        setInitializationStatus(false, false, false, false);
+        assertNotInitializedWithinTwoSeconds();
+
+        setInitializationStatus(true, true, true, true);
+        assertTrue(manager.isInitialized());
+        verify(mockCallback).runWithRetry();
```

Comment:
does this verify that it does it `times(1)`?

Reply:
That's the default, yes -- but can make it explicit